akka-projection-core/src/main/scala/akka/projection/internal/spi/ProjectionMetrics.scala

```diff
@@ -0,0 +1,65 @@
+package akka.projection.internal.spi
+
+import akka.actor.ClassicActorSystemProvider
+import akka.projection.ProjectionId
+
+/**
+ *
+ */
+trait ProjectionMetrics {
+
+  /**
+   * Must be invoked for each element received from the Source when the Source provides it. If possible,
+   * invoke this method in a stream stage as close to the stream stage where you read from the wire to
+   * measure the parsing, deserializing and other steps of the processing prior to the event handling.
+   *
+   * @param projectionId      the projection id
+   * @param creationTimestamp if the element traversing the stream contains the creation time, provide it. Set
+   *                          to null if the information is not available.
+   * @param systemProvider    a `ClassicActorSystemProvider` for telemetry to extract/set data on the ActorSystem
+   * @return a contextual object. This context must propagate with the elt.
+   */
+  def onProcessStart(
+      projectionId: ProjectionId,
+      creationTimestamp: Long,
+      systemProvider: ClassicActorSystemProvider): AnyRef
+
+  /**
+   * Must be invoked for each element processed successfully only when the associated offset has been committed.
+   *
+   * @param projectionId the projection id
+   * @param context      the contextual object returned by `onProcessStart`
+   * @return a contextual object. The returned instance may not be the received one.
+   */
+  def onProcessComplete(projectionId: ProjectionId, context: AnyRef): AnyRef
```

Comment:
what is the returned context supposed to be used for here?

Reply:
why does some of the methods have systemProvider but not all?

Reply:
Cinnamon uses the context to propagate all necessary information from `onProcessStart` to `onProcessComplete`. That context is an arbitrary `Object` where each instrument may attach extra fields. 

`onProcessStart` and `onProcessFailure` get an `ActorSystemProvider` because  those two methods are entry points and will likely need the `ActorSystem` or the `Config` in it to setup any instrument required. `onProcessComplete`, OTOH, will get all necessary information via the context (an instrument may even decide to add a reference to the `ActorSystem` on the context if they need it).

Reply:
Understood, but my question was about the **returned** context from ** onProcessComplete**

Is onProcessFailure an entry point? Looks mor like a result of processing.

Reply:
Sorry, @patriknw I'm having one of those days...

`onProcessComplete` should return `Unit`.

Reply:
> Is onProcessFailure an entry point? Looks mor like a result of processing.

`onProcessFailure` is an entry point in that it is not invoked _after_ any other method invocation. `onProcessStart`/`onProcessComplete` share a context but `onProcessFailure` is independant here (it may even be invoked before `onProcessStart` was ever invoked.

Reply:
Ok, so it’s streamFailure. Not necessarily due to that the event handler failed processing

Reply:
> Ok, so it’s streamFailure. Not necessarily due to that the event handler failed processing

Yes, and again, that's an inherited approach from the Lagom Projection Telemetry SPI. I'm not sure what information could a specific `onProcessFailure` method provide but we could add it and let each instrument decide if they need the ifnromation.

Reply:
If we added `onProcessFailure` we'd be able to report the failure with context information and let the projection developer decide what to do with the failed envelope. In Lagom, a failure in handling caused a stream failure always but maybe we want the extra degree of freedom (maybe a failure processing an envelope is no big deal in some use cases and there's no need to fail the stream and even the offset is stored anyway).